formatter/formatter.go

```diff
@@ -0,0 +1,789 @@
+func (f *formatter) FormatArgumentList(lists ast.ArgumentList) error {
+	f.WriteString("(")
+	for idx, arg := range lists {
+		if err := f.FormatArgument(arg); err != nil {
+			return err
+		}
+
+		if idx != len(lists)-1 {
+			f.NoPadding().WriteWord(",")
+		}
+	}
+	f.WriteString(")")
+
+	return nil
+}
+
+func (f *formatter) FormatArgument(arg *ast.Argument) error {
+	f.WriteWord(arg.Name).NoPadding().WriteString(":").NeedPadding()
+	f.WriteString(arg.Value.String())
+
+	return nil
+}
+
+func (f *formatter) FormatFragmentDefinitionList(lists ast.FragmentDefinitionList) error {
+
+	for _, def := range lists {
+		if err := f.FormatFragmentDefinition(def); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *formatter) FormatFragmentDefinition(def *ast.FragmentDefinition) error {
+
+	f.WriteWord("fragment").WriteWord(def.Name)
+
+	if err := f.FormatVariableDefinitionList(def.VariableDefinition); err != nil {
+		return err
+	}
+
+	f.WriteWord("on").WriteWord(def.TypeCondition)
+
+	if err := f.FormatDirectiveList(def.Directives); err != nil {
+		return err
+	}
+
+	if len(def.SelectionSet) != 0 {
+		if err := f.FormatSelectionSet(def.SelectionSet); err != nil {
+			return err
+		}
+		f.WriteNewline()
+	}
+
+	return nil
+}
+
+func (f *formatter) FormatVariableDefinitionList(lists ast.VariableDefinitionList) error {
+	if len(lists) == 0 {
+		return nil
+	}
+
+	f.WriteString("(")
+	for idx, def := range lists {
+		if err := f.FormatVariableDefinition(def); err != nil {
+			return err
+		}
+
+		if idx != len(lists)-1 {
+			f.NoPadding().WriteWord(",")
+		}
+	}
+	f.NoPadding().WriteString(")").NeedPadding()
+
+	return nil
+}
+
+func (f *formatter) FormatVariableDefinition(def *ast.VariableDefinition) error {
+
+	f.WriteString("$").WriteWord(def.Variable).NoPadding().WriteString(":").NeedPadding()
+	if err := f.FormatType(def.Type); err != nil {
+		return err
+	}
+
+	if def.DefaultValue != nil {
+		f.WriteWord("=")
+		if err := f.FormatValue(def.DefaultValue); err != nil {
+			return err
+		}
+	}
+
+	// TODO https://github.com/vektah/gqlparser/issues/102
+	//   VariableDefinition : Variable : Type DefaultValue? Directives[Const]?
+
+	return nil
+}
+
+func (f *formatter) FormatSelectionSet(sets ast.SelectionSet) error {
+	if len(sets) == 0 {
+		return nil
+	}
+
+	f.WriteString("{").WriteNewline()
+	f.IncrementIndent()
+
+	for _, sel := range sets {
+		if err := f.FormatSelection(sel); err != nil {
+			return err
+		}
+	}
+
+	f.DecrementIndent()
+	f.WriteString("}")
+
+	return nil
+}
+
+func (f *formatter) FormatSelection(selection ast.Selection) error {
+
+	switch v := selection.(type) {
+	case *ast.Field:
+		if err := f.FormatField(v); err != nil {
+			return err
+		}
+
+	case *ast.FragmentSpread:
+		if err := f.FormatFragmentSpread(v); err != nil {
+			return err
+		}
+
+	case *ast.InlineFragment:
+		if err := f.FormatInlineFragment(v); err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("unknown Selection type: %T", selection)
```

Comment:
Is this the only error returned?

Reply:
yes. and, this pattern is not occurred usually.
Are we remove `error` from return values?

Reply:
:+1: yeah lets remove it, will simplify the api and fix your coverage issue.

Reply:
did it!